File: smart-farm-operations-platform-main/frontend/src/types/task.ts
export type TaskStatus = 'TODO' | 'IN_PROGRESS' | 'ON_HOLD' | 'COMPLETED' | 'CANCELLED';
export type TaskPriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export interface FarmTask {
  id: string;
  activityId: string;
  activityTitle: string;
  farmId: string;
  farmName: string;
  fieldId: string;
  fieldName: string;
  title: string;
  description?: string;
  assignedTo?: string;
  assignedToName?: string;
  assignedToPhone?: string;
  assignedBy?: string;
  assignedByName?: string;
  priority: TaskPriority;
  status: TaskStatus;
  dueDate: string;
  completedDate?: string;
  estimatedHours?: number;
  actualHours?: number;
  remarks?: string;
  createdBy: string;
  createdByName?: string;
  createdAt: string;
  updatedAt: string;

  assignedEquipmentId?: string;
  assignedEquipmentName?: string;
  inventoryItemId?: string;
  inventoryQuantityUsed?: number;
  actualCost?: number;
  updatedBy?: string;
  updatedByName?: string;
  gpsLocation?: string;
}

export interface TaskAssignment {
  id: string;
  taskId: string;
  assignedTo?: string;
  assignedBy?: string;
  assignedAt: string;
}

export interface TaskChecklist {
  id: string;
  taskId: string;
  itemName: string;
  checked: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface TaskComment {
  id: string;
  taskId: string;
  commenterId?: string;
  commenterName: string;
  commentText: string;
  createdAt: string;
}

export interface TaskAttachment {
  id: string;
  taskId: string;
  url: string;
  fileName?: string;
  uploadedBy?: string;
  createdAt: string;
}

export interface TaskHistory {
  id: string;
  taskId: string;
  changedBy?: string;
  changedByName?: string;
  previousStatus?: string;
  newStatus?: string;
  remarks?: string;
  createdAt: string;
}

File: smart-farm-operations-platform-main/backend/src/main/java/com/smartfarm/features/task/domain/Task.java
package com.smartfarm.features.task.domain;

import com.smartfarm.features.activity.domain.Activity;
import com.smartfarm.features.farm.domain.Farm;
import com.smartfarm.features.field.domain.Field;
import com.smartfarm.features.auth.domain.User;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;
import org.hibernate.annotations.UpdateTimestamp;

@Entity
@Table(name = "tasks")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@SQLDelete(sql = "UPDATE tasks SET deleted = true, deleted_at = CURRENT_TIMESTAMP WHERE id = ?")
@SQLRestriction("deleted = false")
public class Task {
    @Id
    @GeneratedValue
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "activity_id", nullable = false)
    private Activity activity;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "farm_id", nullable = false)
    private Farm farm;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "field_id", nullable = false)
    private Field field;

    @Column(nullable = false, length = 255)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_to")
    private User assignedTo;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_by")
    private User assignedBy;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 50)
    @Builder.Default
    private TaskPriority priority = TaskPriority.MEDIUM;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 50)
    @Builder.Default
    private TaskStatus status = TaskStatus.TODO;

    @Column(name = "due_date", nullable = false)
    private OffsetDateTime dueDate;

    @Column(name = "completed_date")
    private OffsetDateTime completedDate;

    @Column(name = "estimated_hours", precision = 5, scale = 2)
    private BigDecimal estimatedHours;

    @Column(name = "actual_hours", precision = 5, scale = 2)
    private BigDecimal actualHours;

    @Column(columnDefinition = "TEXT")
    private String remarks;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_equipment_id")
    private com.smartfarm.features.operations.domain.Equipment assignedEquipment;

    @Column(name = "inventory_item_id")
    private UUID inventoryItemId;

    @Column(name = "inventory_quantity_used", precision = 10, scale = 2)
    private BigDecimal inventoryQuantityUsed;

    @Column(name = "actual_cost", precision = 12, scale = 2)
    private BigDecimal actualCost;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by")
    private User createdBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "updated_by")
    private User updatedBy;

    @Column(name = "gps_location")
    private String gpsLocation;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @Builder.Default
    private boolean deleted = false;

    @Column(name = "deleted_at")
    private OffsetDateTime deletedAt;
}
